src/rust/src/wallet.rs

```diff
@@ -0,0 +1,428 @@
+    wallet.rewind_witness_tree()
+}
+
+#[no_mangle]
+pub extern "C" fn orchard_wallet_add_notes_from_bundle(
+    wallet: *mut Wallet,
+    txid: *const [c_uchar; 32],
+    bundle: *const Bundle<Authorized, Amount>,
+) {
+    let wallet = unsafe { wallet.as_mut() }.expect("Wallet pointer may not be null");
+    let txid = TxId::from_bytes(*unsafe { txid.as_ref() }.expect("txid may not be null."));
+    if let Some(bundle) = unsafe { bundle.as_ref() } {
+        wallet.add_notes_from_bundle(&txid, bundle);
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn orchard_wallet_append_bundle_commitments(
+    wallet: *mut Wallet,
+    block_height: u32,
+    block_tx_idx: usize,
+    txid: *const [c_uchar; 32],
+    bundle: *const Bundle<Authorized, Amount>,
+) -> bool {
+    let wallet = unsafe { wallet.as_mut() }.expect("Wallet pointer may not be null");
+    let txid = TxId::from_bytes(*unsafe { txid.as_ref() }.expect("txid may not be null."));
+    if let Some(bundle) = unsafe { bundle.as_ref() } {
+        if let Err(e) =
+            wallet.append_bundle_commitments(block_height.into(), block_tx_idx, &txid, bundle)
+        {
+            error!("An error occurred adding the Orchard bundle's notes to the note commitment tree: {:?}", e);
+            return false;
+        }
+    }
+
+    true
+}
+
+#[no_mangle]
+pub extern "C" fn orchard_wallet_add_spending_key(wallet: *mut Wallet, sk: *const SpendingKey) {
+    let wallet = unsafe { wallet.as_mut() }.expect("Wallet pointer may not be null");
+    let sk = unsafe { sk.as_ref() }.expect("Spending key may not be null.");
+
+    wallet.key_store.add_spending_key(*sk);
+}
+
+#[no_mangle]
+pub extern "C" fn orchard_wallet_add_full_viewing_key(
+    wallet: *mut Wallet,
+    fvk: *const FullViewingKey,
+) {
+    let wallet = unsafe { wallet.as_mut() }.expect("Wallet pointer may not be null.");
+    let fvk = unsafe { fvk.as_ref() }.expect("Full viewing key pointer may not be null.");
+
+    wallet.key_store.add_full_viewing_key(fvk.clone());
+}
+
+#[no_mangle]
+pub extern "C" fn orchard_wallet_add_raw_address(
+    wallet: *mut Wallet,
+    addr: *const Address,
+    ivk: *const IncomingViewingKey,
+) -> bool {
+    let wallet = unsafe { wallet.as_mut() }.expect("Wallet pointer may not be null.");
+    let addr = unsafe { addr.as_ref() }.expect("Address may not be null.");
+    let ivk = unsafe { ivk.as_ref() }.expect("Incoming viewing key may not be null.");
+
+    wallet.key_store.add_raw_address(*addr, ivk.clone())
+}
+
+#[no_mangle]
+pub extern "C" fn orchard_wallet_get_ivk_for_address(
+    wallet: *const Wallet,
+    addr: *const Address,
+) -> *mut IncomingViewingKey {
+    let wallet = unsafe { wallet.as_ref() }.expect("Wallet pointer may not be null.");
+    let addr = unsafe { addr.as_ref() }.expect("Address may not be null.");
+
+    wallet
+        .key_store
+        .ivk_for_address(addr)
+        .map(|ivk| Box::into_raw(Box::new(ivk.clone())))
```

Comment:
It's unclear how this memory is freed.

Reply:
It's freed in the same fashion as other means of obtaining an Orchard incoming viewing key, using `orchard_incoming_viewing_key_free` (which was added in a much earlier commit.)